kernel/src/storage_permissions.rs

```diff
@@ -53,29 +61,76 @@ impl StoragePermissions {
             write_count: write_count_capped,
             write_permissions,
             write_id,
+            superuser: false,
+        }
+    }
+
+    /// Create superuser permissions suitable for the kernel. This allows the
+    /// kernel to read/update any stored item, and allows the kernel to write
+    /// items that will not be accessible to any clients without superuser
+    /// permissions.
+    pub fn new_kernel_permissions(_cap: &dyn capabilities::SuperuserStorageCapability) -> Self {
+        let read_permissions: [u32; 8] = [0; 8];
+        let write_permissions: [u32; 8] = [0; 8];
+        StoragePermissions {
+            read_count: 0,
+            read_permissions,
+            write_count: 0,
+            write_permissions,
+            write_id: None,
+            superuser: true,
         }
     }
 
     /// Check if this permission object grants read access to the specified
     /// `storage_id`. Returns `true` if access is permitted, `false` otherwise.
     pub fn check_read_permission(&self, storage_id: u32) -> bool {
-        self.read_permissions
-            .get(0..self.read_count)
-            .unwrap_or(&[])
-            .contains(&storage_id)
+        if self.superuser {
+            // Superuser grants all permissions.
+            true
```

Comment:
I have a few thoughts about this.

Obviously the kernel can do anything, it has direct access to the flash. But I think it would make more sense to allow apps to specify if the kernel can access their files. Accesses happen from capsules, which are only semi-trusted in Tock. That allows applications to apply some restrictions on what the kernel can/can't do.

I also worry about having a single branch instruction to allow/disallow access like this. Doesn't that expose us to glitch attacks? In saying that though, I'm not sure if anything else here is hardened against attacks like that.

I think it would be better to make the kernel ID `0x00` a user, the same as all other users. That way apps can choose to share data with the kernel or not.

Reply:
That seems reasonable to me. I think my current use case doesn't need superuser permissions, just having reserved kernel storage is sufficient. Updated.